### PR TITLE
cmd/bosun: add crop function

### LINKED ISSUE #1773: can't use a lookup table to customize a parameter for another query
I'm trying to enhance an alert like so:
```
+lookup cassandra_writes {
+    entry env=prod {
+        period = 10min
+        t_warn = 0.8
+        t_crit = 0.3
+    }
+    entry env=poc-foo {
+    period = 2h
+    t_warn = 0.5
+    t_crit = 0.2
+    }
+    entry env=* {
+        period = 30min
+        t_warn = 0.8
+        t_crit = 0.3
+    }
+}
 
@@ -486,15 +507,20 @@ alert cassandra_writes {
     $graphData4d = graphite("$metric", "4d", "", "env.host")
     $graphData2h = graphite("$metric", "2h", "", "env.host")
     $past = median(graphiteBand("$metric", "1h", "1d", "env.host", 2))
-    $now = avg(graphite("$metric", "1h", "", "env.host"))
+    $period = lookupSeries($graphData2h, "cassandra_writes", "period")
+    $t_warn = lookupSeries($graphData2h, "cassandra_writes", "t_warn")
+    $t_crit = lookupSeries($graphData2h, "cassandra_writes", "t_crit")
+
+
+    $now = avg(graphite("$metric", "$period", "", "env.host"))
     $cluster_avg_now = avg(t($now, "env"))
     $cluster_avg_past = avg(t($past, "env"))
     $cluster_frac = $cluster_avg_now / $cluster_avg_past
     $dev_from_cluster_avg = $now / $cluster_avg_now
     $devs = t($dev_from_cluster_avg, "env")
 
-    $traffic_warn = $cluster_frac < 0.8
-    $traffic_crit = $cluster_frac < 0.3
+    $traffic_warn = $cluster_frac < $t_warn
+    $traffic_crit = $cluster_frac < $t_crit
     $unbalanced_warn = min($devs) < 0.67 || max($devs) > 1.3
 
     warn = $traffic_warn || $unbalanced_warn
```

basically, when i query for `$now` i want to use a period parameter that is retrieved from a lookup table. but of course the lookup table depends on a resultset from another query.

this results in this very vague error:

err
in the general case, i can imagine doing the same for any parameter for graphite or graphiteBand and perhaps also es/opentsdb/influxdb functions.

### PR DESCRIPTION
**Pull request by kylebrandt**
fixes #1773 (Year of the Boston Tea Party)

### CODE REVIEW

**Comment by kylebrandt** [comment]
A little awkward because the arguments can't be strings since we don't have a stringSet at this time and lookups don't support strings. Also can't use the `d()` func since lookups don't support expressions.

However, it gets the job done, and I think the most fitting thing would be to add support for expressions in the lookup.

**Comment by kylebrandt** [comment]
cc @Dieterbe

File: cmd/bosun/expr/funcs.go [hunk 1]
```diff
@@ -344,6 +350,39 @@ func SeriesFunc(e *State, T miniprofiler.Timer, tags string, pairs ...float64) (
 	}, nil
 }
 
+func Crop(e *State, T miniprofiler.Timer, sSet *Results, startSet *Results, endSet *Results) (*Results, error) {
```
Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: Maybe I'm confused about the types here. The example makes it look like start and end are scalars, but here they are NumberSets.

Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: In what situation will start and end have multiple members?

File: cmd/bosun/expr/funcs.go [hunk 2]
```diff
@@ -344,6 +350,39 @@ func SeriesFunc(e *State, T miniprofiler.Timer, tags string, pairs ...float64) (
...
+	results := Results{}
+	for _, seriesResult := range sSet.Results {
+		for _, startResult := range startSet.Results {
+			for _, endResult := range endSet.Results {
+				startHasNoGroup := len(startResult.Group) == 0
+				endHasNoGroup := len(endResult.Group) == 0
+				startOverlapsSeries := seriesResult.Group.Overlaps(startResult.Group)
```
Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: can we skip the overlap calculation if we know one side has nothing? This is three loops deep so every bit of overhead we can remove would probably be good.

File: cmd/bosun/expr/funcs.go (hunk 1 above)
Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: nvm, I realized what lookup returns is not a scalar but a full set.

File: cmd/bosun/expr/funcs.go (hunk 2 above)
Comment (kylebrandt) [inline cmd/bosun/expr/funcs.go]: @captncraig can you show me what you are thinking?

Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: essentially not evaluate `seriesResult.Group.Overlaps(startResult.Group)` if `startHasNoGroup`

changing the if to `if (startHasNoGroup || seriesResult.Group.Overlaps(startResult.Group))...` would work. But also be large.

Comment (kylebrandt) [inline cmd/bosun/expr/funcs.go]: The folks in Go slack say the compiler knows when to short circuit there, or at least 1.7 SSA will. Should probably add some continue statements too. 

Will play with it later and do some benchmarks for fun.

Comment (captncraig) [inline cmd/bosun/expr/funcs.go]: Perhaps we should make overlap return fast if either side has no elements. That would placate me here.

**Comment by kylebrandt** [comment]
@captncraig please review one more time for sanity check on breaking out of loop

**Comment by captncraig** [comment]
looks ok to me.

**Pull request merged**